File: IDATT2106 Systemutvikling 2 med smidig prosjekt/backend/src/test/java/com/ntnu/gidd/factories/UserFactory.java
package com.ntnu.gidd.factories;

import com.ntnu.gidd.model.User;
import org.springframework.beans.factory.FactoryBean;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;

import java.time.LocalDate;
import java.time.ZonedDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.UUID;

import static com.ntnu.gidd.utils.StringRandomizer.getRandomEmail;
import static com.ntnu.gidd.utils.StringRandomizer.getRandomString;

public class UserFactory implements FactoryBean<User> {

    private BCryptPasswordEncoder encoder = new BCryptPasswordEncoder();


    @Override
    public User getObject() throws Exception {
        return User.builder()
                .id(UUID.randomUUID())
                .email(getRandomEmail())
                .firstName(getRandomString(5))
                .surname(getRandomString(5))
                .password(encoder.encode(getRandomString(10)))
                .following(new ArrayList<>())
                .followers(new ArrayList<>())
                .birthDate(LocalDate.now())
                .createdAt(ZonedDateTime.now())
                .build();
    }

    @Override
    public Class<?> getObjectType() {
        return null;
    }

    @Override
    public boolean isSingleton() {
        return false;
    }
}

File: IDATT2105 Full-stack applikasjonsutvikling/Rombestilling/backend/src/test/kotlin/ntnu/idatt2105/factories/UserFactory.kt
package ntnu.idatt2105.factories

import io.github.serpro69.kfaker.Faker
import io.github.serpro69.kfaker.FakerConfig
import io.github.serpro69.kfaker.create
import ntnu.idatt2105.user.model.RoleType
import ntnu.idatt2105.user.model.User
import ntnu.idatt2105.user.model.UserBuilder
import org.springframework.beans.factory.FactoryBean
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
import java.lang.Exception
import java.time.LocalDate
import java.util.*

class UserFactory : FactoryBean<User> {

    private val config = FakerConfig.builder().create {
        uniqueGeneratorRetryLimit = 10000
    }

    private val faker = Faker(config)

    private val encoder = BCryptPasswordEncoder()

    @Throws(Exception::class)
    override fun getObject(): User {
        faker.unique.clearAll()
        faker.unique.configuration { enable(faker::internet) }
        faker.unique.configuration { enable(faker::name) }
        val userRole = RoleFactory().`object`

        return UserBuilder(
            id = UUID.randomUUID(),
            email = faker.internet.email(),
            firstName = faker.name.firstName(),
            surname = faker.name.lastName(),
            password = encoder.encode(faker.rickAndMorty.locations()),
            phoneNumber = faker.phoneNumber.phoneNumber(),
            expirationDate = LocalDate.now().plusDays(1),
            roles = mutableSetOf(userRole)
        ).build()
    }

    fun admin(): User =
        this.`object`.copy(roles = mutableSetOf(
            RoleFactory().`object`,
            RoleFactory().getObject(RoleType.ADMIN)
        ))

    override fun getObjectType(): Class<*>? {
        return null
    }

    override fun isSingleton(): Boolean {
        return false
    }
}
